Create linalg.tensordot

antocuni opened the pull request:
PR stack (ghstack-style but done manually)
* #64819
* __->__  #63478

Fixes #61649.
As the title says, this introduces `torch.linalg.tensordot` and makes `torch.tensordot` an alias to it.
However, `tensordot` is a bit different than most other `linalg` operators so it is worth some discussion on how to proceed.
Differently than the other operators, the user entry-point of `tensordot` is implemented in python, in `torch/functional.py`, which in turns calls `torch._C._linalg.linalg_tensordot`.

What I did so far in this PR is:
1. introduce the C++-level `at::linalg_tensordot` and make it the main implementation
2. make `at::tensordot` a C++ alias to `at::linalg_tensordot`
3. leave the python implementation inside `torch/functional.py`
4. make it available as both `torch.tensordot` and `torch.linalg.tensordot`
5. update all the docs to use `torch.linalg.tensordot` everywhere
6. update all the existing code to use `torch.linalg.tensordot` everywhere

I am not fully sure about point (3) though. One possible alternative is to move the implementation to e.g. a newly created `torch/linalg/functional.py`, but this creates more problems because there are other pieces of code and/or tests which expect all these functions to be inside `torch/functional.py`.

The other problem of the current solution is that since `torch.tensordot` and `torch.linalg.tensordot` are the very same python-level function, they have the very same docstring and thus the same generated documentation page. Maybe it is possible to tweak sphinx to use a custom text for `torch.tensordot` instead of using the docstring, but I haven't investigate yet.

I opted for the current solution because it was the easiest to implement and I wanted to get some feedback before spending time in more complex ones. Do you think it is enough or we should try to differentiate more between `torch.tensordot` and `torch.linalg.tensordot`?


cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

antocuni committed:
introduce linalg.tensordot and make torch.tensordot an alias to it

antocuni committed:
replace torch.tensordot with torch.linalg.tensordot everywhere

antocuni committed:
make linalg.tensordot the main op and torch.tensordot an alias to it in common_method_invocations

antocuni committed:
fix mypy

antocuni committed:
intern this string

antocuni committed:
fix docs, and use torch.linalg.tensordot everywhere

antocuni committed:
fix ONNX

antocuni committed:
fix tensordot autocast

IvanYashchuk commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4238,13 +4238,12 @@
   dispatch:
     CPU, CUDA: tanh_out
 
+# Alias to linalg.tensordot
```
Adding an alias requires a few steps described in `# Note [Adding an alias]` in the same file `native_functions.yaml`.
Step 6 is missing.

antocuni replied:
thanks, fixed by 7e7cb83239

IvanYashchuk commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10514,6 +10513,16 @@
+  python_module: linalg
+  variants: function
+
+- func: linalg_tensordot.out(Tensor self, Tensor other, int[] dims_self, int[] dims_other, *, Tensor(a!) out) -> Tensor(a!)
```
I think it's worth investing the time and port the implementation to be a "structured kernel". See https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#structured-keyword-proposal

IvanYashchuk commented on torch/functional.py:
```diff
@@ -886,22 +887,22 @@ def tensordot(a, b, dims=2, out: Optional[torch.Tensor] = None):  # noqa: F811
 
     When called with a non-negative integer argument :attr:`dims` = :math:`d`, and
     the number of dimensions of :attr:`a` and :attr:`b` is :math:`m` and :math:`n`,
-    respectively, :func:`~torch.tensordot` computes
+    respectively, :func:`torch.linalg.tensordot` computes
 
     .. math::
         r_{i_0,...,i_{m-d}, i_d,...,i_n}
           = \sum_{k_0,...,k_{d-1}} a_{i_0,...,i_{m-d},k_0,...,k_{d-1}} \times b_{k_0,...,k_{d-1}, i_d,...,i_n}.
 
     When called with :attr:`dims` of the list form, the given dimensions will be contracted
     in place of the last :math:`d` of :attr:`a` and the first :math:`d` of :math:`b`. The sizes
-    in these dimensions must match, but :func:`~torch.tensordot` will deal with broadcasted
+    in these dimensions must match, but :func:`torch.linalg.tensordot` will deal with broadcasted
     dimensions.
```
Documentation for `tensordot` and implementation should be moved to `linalg/__init__.py`. Leaving here only `tensordot = torch.linalg.tensordot`.

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -2036,3 +2036,5 @@
     tensor(6.2158e-07)
 """)
+
+tensordot = torch.functional.tensordot
```
`torch.linalg.tensordot` should be the primary function and `torch.functional.tensordot` use it.
Documentation for `torch.linalg.tensordot` should be in this file like for any other function from `linalg` module.

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -7936,30 +7936,30 @@ def fn(torchfn, *args):
     def test_tensordot(self, device):
```
Is this test good enough and doesn't need to be updated? Are all supported types of `dims` tested here?

IvanYashchuk replied:
I think at least dtype parametrization should be added.

antocuni replied:
good idea! Actually all the various possibilities were already tested by `test_jit.py` but it looks like a good idea to have a more direct test here. Fixed by 053beac580 and e7fc1ea4b1

IvanYashchuk commented on torch/__init__.py:
```diff
@@ -622,7 +622,6 @@ def manager_path():
 # needs to be after the above ATen bindings so we can overwrite from Python side
 from .functional import *  # noqa: F403
 
-
```
Touching this file is probably unnecessary.

IvanYashchuk commented on torch/functional.py:
```diff
@@ -13,6 +13,7 @@
 from torch import _VF
+from torch._C import _linalg  # type: ignore[attr-defined]
 
 __all__ = [
```
Removing `tensordot` from `__all__` would decouple `torch.tensordot` and `torch.functional.tensordot`. Then you'd get it as an alias to `torch._C._linalg.linalg_tensordot` and it should be possible to add a separate docstring in `_torch_docs.py` that would be simply referring to `torch.linalg.tensordot`.

antocuni replied:
I don't think this is doable: if I make `torch.tensordot` an alias to `torch._C._linalg.linalg_tensordot` it would have a different signature and behavior as `torch.linalg.tensordot`, because the latter has a python implementation.

IvanYashchuk commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10514,6 +10513,16 @@
 - func: linalg_multi_dot.out(Tensor[] tensors, *, Tensor(a!) out) -> Tensor(a!)
   python_module: linalg
 
+- func: linalg_tensordot(Tensor self, Tensor other, int[] dims_self, int[] dims_other) -> Tensor
```
The signature in Python is `tensordot(a, b, dims=2, out=None)` where `dims (int or Tuple[List[int], List[int]] or List[List[int]] containing two lists or Tensor)`. Can all the logic on `dims` argument be ported from Python to C++? If not maybe the Python signature should be modified so that the mapping from Python to C++ is more direct and no glue code from functional.py is neccessary?

antocuni replied:
honestly, I don't know the answer to this question. The fact that the logic was originally written in Python makes me to suspect that there was a good reason to do that but I don't really know if the original reason still holds.
E.g., maybe the dispatcher cannot handle such a complex overload?

Lezcano commented on docs/source/linalg.rst:
```diff
@@ -80,6 +80,7 @@ Tensor Operations
 
     tensorinv
     tensorsolve
+    tensordot
```
Nit. Probably put it before `tensorinv` and `tensorsolve` as it's the main operation

antocuni replied:
done af85b06e0b

Lezcano commented on docs/source/linalg.rst:
```diff
@@ -80,6 +80,7 @@ Tensor Operations
 
     tensorinv
     tensorsolve
```
It'd be good also to update the docs for `tensorinv` and `tensorsolve` to point to `linalg.tensordot` rather than `torch.tensordot`.

antocuni replied:
I'm confused: isn't it exactly what I did in commit 0d17d668f9?

Lezcano replied:
Ah, yes, sorry, I missed that!

antocuni committed:
add the new alias to the alias map

antocuni committed:
move the implementation of tensordot from functional.py to linalg/__init__.py

antocuni committed:
tentatively try to teach the JIT about the new linalg.tensordot operation

antocuni committed:
torch.functional.tensordot is explictly listed in the torchscript docs. Make sure to preserve compatibility

antocuni committed:
Merge branch 'master' into antocuni/linalg-tensordot

antocuni committed:
reduce the diff w.r.t. master

antocuni committed:
put tensordot first

antocuni committed:
kill leftover

antocuni committed:
rename test_functional_tensordot_* into test_linalg_tensordot_*

antocuni committed:
improve test_tensordot by testing also the variants in which dims is a Tensor

antocuni committed:
improve the test to use multiple dtypes

antocuni committed:
fix test_fx

antocuni committed:
fix flake8

antocuni committed:
fix mypy

antocuni committed:
torch.linalg imports _overload as overload so it looks like a public function, but actually it's not

antocuni committed:
Merge branch 'master' into antocuni/linalg-tensordot

antocuni committed:
fix

antocuni committed:
fix flake8

antocuni commented:
@IvanYashchuk @Lezcano I think I addressed all your comments and I fixed the conflicts with master.
The only two comments which I didn't address are the ones about using a structured kernel and move more logic into C++.

I already started to look into it but I'd prefer to do it in a separate PR to keep this one manageable and avoid more conflicts.

antocuni committed:
try to fix flake8 again

antocuni committed:
Merge branch 'master' into antocuni/linalg-tensordot

mruberry commented on torch/__init__.py:
```diff
@@ -769,6 +770,11 @@ def compiled_with_cxx11_abi():
+# this was originally defined in torch.functional and made available as
+# torch.tensordot. Now is has been moved to torch.linalg.tensordot, but we
+# keep it available in the old namespaces to preserve backward compatibility.
+tensordot = linalg.tensordot
```
Can't the functional tensordot stick around and just call the function it always has though instead of changing __init__.py?

mruberry commented on torch/jit/_builtins.py:
```diff
@@ -117,13 +117,13 @@ def _gen_torch_functional_registered_ops():
 
 def _is_special_functional_bound_op(fn):
-    return fn in _functional_registered_ops
+    return fn in _functional_registered_ops or fn is torch.linalg.tensordot
```
This kind of special-casing is worrisome

mruberry commented:
Creating a linalg.tensordot alias is unfortunately tricky because of all the special handling we have for functionals. 

To add it we'd probably need to create linalg.functional.py and then also bind the function defined there in functional.py? 

There are some other issues with tensordot itself that may be interesting to address first:

- making it structured
- correctness 
- docs

I filed https://github.com/pytorch/pytorch/issues/65989 for the correctness issue (and a suggestion to fix the docs). PRs addressing these three issues would be extremely interesting if you'd prefer to start with those, @antocuni. 

For the docs in particular it'd be helpful to borrow a page from NumPy and show the common cases as well as their equivalent computations. The current torch docs are nearly indecipherable because they use undefined terms and the mathematical formula is (at best) very obtuse.

Lezcano commented on torch/fx/operator_schemas.py:
```diff
@@ -253,7 +253,10 @@ def normalize_function(
     new_args_and_kwargs = None
-    if target in boolean_dispatched or target.__module__ in ['torch.nn.functional', 'torch.functional']:
+
+    if ((target in boolean_dispatched or
```
I have not been following the discussion, but why would we want `torch.linalg.functional`? Couldn't we just port all this to C++ as @IvanYashchuk mentioned in https://github.com/pytorch/pytorch/pull/63478/files/e12028da72240f2e97f25c4ad01640f1655f5b1f#r701166925 ?

mruberry replied:
Implementing it entirely in C++ would be ideal but I think the issue with that is that we don't support list of lists in native_functions.yaml? 

I think each item in native_functions.yaml must be a base type or a list of a base type, and the base types are defined here:

https://github.com/pytorch/pytorch/blob/ea0de37d2e5aa71178580f3bce7559cc6a3b9135/tools/codegen/model.py#L948

(they may also be redundantly defined elsewhere but this is the thing that does the codegen)

antocuni committed:
undo 2380bfa90f and make torch.functional.tensordot available more directly

antocuni committed:
create torch.linalg.functional and move tensordot there

antocuni committed:
Merge branch 'master' into antocuni/linalg-tensordot

mruberry reviewed (commented):
This looks really good, @antocuni. I made an inline comment about preserving BC. There are a few other places that use tensordot that may need an update:

https://github.com/pytorch/pytorch/blob/master/docs/source/amp.rst

and

https://github.com/pytorch/pytorch/blob/713e025c9f43a5dd50c25a2055b9e120b40398e2/torch/testing/_internal/autocast_test_lists.py#L199

(we should be sure both tensordot and linalg.tensordot are handled properly)

and the last change I'd like to suggest is that while linalg.tensordot is the suggested user-facing op that internally we don't change the implementation of tensordot and actually alias linalg.tensordot to it. This might avoid some internal refactoring issues (that we don't expect to apply in the future).

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-tensordot

antocuni committed:
readd these for BC

antocuni committed:
move tensordot back to torch.functional

antocuni committed:
update docs

antocuni committed:
add an autocast promote test for linalg.tensordot

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -9712,7 +9712,7 @@ def forward(self, x, window_length: int):
         class M(torch.nn.Module):
             def forward(self, x, y):
-                output = torch.tensordot(x, y, 2)
+                output = torch.linalg.tensordot(x, y, 2)
```
Should we keep one redundant test for `torch.tensordot()`?

mruberry commented on test/test_linalg.py:
```diff
@@ -8029,38 +8029,56 @@ def fn(torchfn, *args):
-        b = torch.randn(4, 5, 6, 7, device=device)
-        c = torch.tensordot(a, b, dims=2).cpu()
+        # case 4: dims is an integer
+        a = torch.randn(2, 3, 4, 5, dtype=dtype, device=device)
```
Nice dtype updates

mruberry commented on test/test_linalg.py:
```diff
@@ -8029,38 +8029,56 @@ def fn(torchfn, *args):
         self.assertRaises(RuntimeError, lambda: torch.lstsq(torch.randn(0, 0), torch.randn(0, 0)))
         self.assertRaises(RuntimeError, lambda: torch.lstsq(torch.randn(0,), torch.randn(0, 0)))
 
+    @dtypes(*floating_types())
```
We can probably just run this in tf32 still unless there's a particular reason this would like to test double, too?

antocuni replied:
yes, there is no particular reason. I switched to testing only `torch.float32`.

antocuni replied:
yes, there is no particular reason. I switched to testing only torch.float32 in 33345f7cdf

mruberry commented on test/test_linalg.py:
```diff
@@ -8029,38 +8029,56 @@ def fn(torchfn, *args):
 
-        a = torch.tensordot(torch.tensor(0.), torch.tensor(0.), 0)
+        # case 8: dims=0
+        a = torch.linalg.tensordot(torch.tensor(0.), torch.tensor(0.), 0)
```
These tensors need to use the device and dtype, too

antocuni replied:
right, added in d62803a2a6

mruberry commented on torch/functional.py:
```diff
@@ -1156,9 +1156,10 @@ def tensordot(a, b, dims=2, out: Optional[torch.Tensor] = None):  # noqa: F811
 
     if out is None:
-        return _VF.tensordot(a, b, dims_a, dims_b)  # type: ignore[attr-defined]
+        return torch._C._linalg.linalg_tensordot(a, b, dims_a, dims_b)  # type: ignore[attr-defined]
```
I'm a little confused by this part still. This functional has a bunch of logic before calling `torch._C._linalg.linalg_tensordot()`, but when `torch.linalg.tensordot()` is called how does it implement all the above logic?

antocuni replied:
`torch.linalg.tensordot` is the very same function as `torch.functional.tensordot`:
```
>>> import torch
>>> torch.linalg.tensordot is torch.functional.tensordot
True
```

mruberry reviewed (commented):
Hey @antocuni! Sorry to make you wait and thanks for your patience. I'm finally in the NYC office.

Overall this looks great; I added just a few inline comments for your review. Looking forward to hearing your thoughts!

antocuni committed:
no need to test float64

antocuni committed:
add dtype and device also to this tensor

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/linalg-tensordot

antocuni commented:
@mruberry hopefully addressed your latest comments. I also updated to the latest master and fixed conflicts (tests still running at the time of writing, hopefully they will be green)
